src/algebra/pointwise.lean

```diff
@@ -880,10 +880,35 @@ by { simp only [← image2_mul, image_image2, image2_image_left, image2_image_ri
 lemma preimage_mul_preimage_subset {s t : set β} : m ⁻¹' s * m ⁻¹' t ⊆ m ⁻¹' (s * t) :=
 by { rintros _ ⟨_, _, _, _, rfl⟩, exact ⟨_, _, ‹_›, ‹_›, (m.map_mul _ _).symm ⟩ }
 
+instance set_semiring.no_zero_divisors : no_zero_divisors (set_semiring α) :=
+{ eq_zero_or_eq_zero_of_mul_eq_zero := λ a b ab, by
+  { by_cases a0 : a = 0,
+  { exact or.inl a0 },
+  { refine or.inr _,
+    by_cases b0 : b = 0,
+    { exact b0 },
+    { cases ne_empty_iff_nonempty.mp a0 with x xa,
+      cases ne_empty_iff_nonempty.mp b0 with y yb,
+      exact (not_not.mpr ab (ne_empty_iff_nonempty.mpr ⟨x * y, mul_mem_mul xa yb⟩)).elim } } } }
+
+/- Since addition on `set_semiring` is commutative (it is set union), there is no need
+to also have the instance `covariant_class (set_semiring α) (set_semiring α) (swap (+)) (≤)`. -/
+instance set_semiring.covariant_class_add :
+  covariant_class (set_semiring α) (set_semiring α) (+) (≤) :=
+{ elim := λ a b c ab x xa, mem_of_mem_of_subset xa (union_subset_union_right _ ab) }
+
+instance set_semiring.covariant_class_mul_left :
+  covariant_class (set_semiring α) (set_semiring α) (*) (≤) :=
+{ elim := λ a b c ab x xa, mem_of_mem_of_subset xa (mul_subset_mul subset.rfl ab) }
+
+instance set_semiring.covariant_class_mul_right :
+  covariant_class (set_semiring α) (set_semiring α) (swap (*)) (≤) :=
+{ elim := λ a b c ab x xa, mem_of_mem_of_subset xa (mul_subset_mul ab subset.rfl) }
+
 end mul_hom
 
-/-- The image of a set under function is a ring homomorphism
-with respect to the pointwise operations on sets. -/
+/-- The image of a set under a multiplicative homomorphism is a ring homomorphism
+with respect to the pointwise operations on subsets. -/
```

Comment:
I think `sets` was more descriptive. I hope we don't try to adhere to the confusing ZFC nomenclature!

Reply:
I accepted your suggestion, but in my mind I change `set` to `subsets`!